// yt-transcript-api/src/errors.js
class YouTubeTranscriptApiException extends Error { }

class CookieError extends YouTubeTranscriptApiException { }

class CookiePathInvalid extends CookieError {
   constructor(cookiePath) {
      super(`Can't load the provided cookie file: ${cookiePath}`);
   }
}

class CookieInvalid extends CookieError {
   constructor(cookiePath) {
      super(`The cookies provided are not valid (may have expired): ${cookiePath}`);
   }
}

class CouldNotRetrieveTranscript extends YouTubeTranscriptApiException {
   static ERROR_MESSAGE = "\nCould not retrieve a transcript for the video {video_url}!";
   static CAUSE_MESSAGE_INTRO = " This is most likely caused by:\n\n{cause}";
   static CAUSE_MESSAGE = "";
   static GITHUB_REFERRAL = (
      "\n\nIf you are sure that the described cause is not responsible for this error "
      + "and that a transcript should be retrievable, please create an issue at "
      + "https://github.com/chintamani-pala/yt-transcript-api/issues. "
      + "Please add which version of yt-transcript-api you are using "
      + "and provide the information needed to replicate the error. "
      + "Also make sure that there are no open issues which already describe your problem!"
   );

   constructor(videoId) {
      super();
      this.videoId = videoId;
   }

   _buildErrorMessage() {
      // You must import WATCH_URL from settings.js in your project
      const { WATCH_URL } = require('./settings');
      let errorMessage = CouldNotRetrieveTranscript.ERROR_MESSAGE.replace(
         '{video_url}',
         WATCH_URL.replace('{video_id}', this.videoId)
      );
      const cause = this.cause;
      if (cause) {
         errorMessage += (
            CouldNotRetrieveTranscript.CAUSE_MESSAGE_INTRO.replace('{cause}', cause)
            + CouldNotRetrieveTranscript.GITHUB_REFERRAL
         );
      }
      return errorMessage;
   }

   get cause() {
      return CouldNotRetrieveTranscript.CAUSE_MESSAGE;
   }

   toString() {
      return this._buildErrorMessage();
   }
}

class YouTubeDataUnparsable extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = (
      "The data required to fetch the transcript is not parsable. This should "
      + "not happen, please open an issue (make sure to include the video ID)!"
   );
   get cause() {
      return YouTubeDataUnparsable.CAUSE_MESSAGE;
   }
}

class YouTubeRequestFailed extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "Request to YouTube failed: {reason}";
   constructor(videoId, httpError) {
      super(videoId);
      this.reason = String(httpError);
   }
   get cause() {
      return YouTubeRequestFailed.CAUSE_MESSAGE.replace('{reason}', this.reason);
   }
}

class VideoUnplayable extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "The video is unplayable for the following reason: {reason}";
   static SUBREASON_MESSAGE = "\n\nAdditional Details:\n{sub_reasons}";
   constructor(videoId, reason, subReasons) {
      super(videoId);
      this.reason = reason;
      this.subReasons = subReasons;
   }
   get cause() {
      let reason = this.reason == null ? "No reason specified!" : this.reason;
      if (this.subReasons && this.subReasons.length > 0) {
         const subReasons = this.subReasons.map(sub => ` - ${sub}`).join('\n');
         reason = `${reason}${VideoUnplayable.SUBREASON_MESSAGE.replace('{sub_reasons}', subReasons)}`;
      }
      return VideoUnplayable.CAUSE_MESSAGE.replace('{reason}', reason);
   }
}

class VideoUnavailable extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "The video is no longer available";
   get cause() {
      return VideoUnavailable.CAUSE_MESSAGE;
   }
}

class InvalidVideoId extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = (
      "You provided an invalid video id. Make sure you are using the video id and NOT the url!\n\n"
      + 'Do NOT run: `YouTubeTranscriptApi.get_transcript("https://www.youtube.com/watch?v=1234")`\n'
      + 'Instead run: `YouTubeTranscriptApi.get_transcript("1234")`'
   );
   get cause() {
      return InvalidVideoId.CAUSE_MESSAGE;
   }
}

class RequestBlocked extends CouldNotRetrieveTranscript {
   static BASE_CAUSE_MESSAGE = (
      "YouTube is blocking requests from your IP. This usually is due to one of the "
      + "following reasons:\n"
      + "- You have done too many requests and your IP has been blocked by YouTube\n"
      + "- You are doing requests from an IP belonging to a cloud provider (like AWS, "
      + "Google Cloud Platform, Azure, etc.). Unfortunately, most IPs from cloud "
      + "providers are blocked by YouTube.\n\n"
   );
   static CAUSE_MESSAGE = (
      RequestBlocked.BASE_CAUSE_MESSAGE
      + "There are two things you can do to work around this:\n"
      + '1. Use proxies to hide your IP address, as explained in the "Working around '
      + 'IP bans" section of the README '
      + "(https://github.com/chintamani-pala/yt-transcript-api"
      + "\n"
      + "2. (NOT RECOMMENDED) If you authenticate your requests using cookies, you "
      + "will be able to continue doing requests for a while. However, YouTube will "
      + "eventually permanently ban the account that you have used to authenticate "
      + "with! So only do this if you don't mind your account being banned!"
   );
   static WITH_GENERIC_PROXY_CAUSE_MESSAGE = (
      "YouTube is blocking your requests, despite you using proxies. Keep in mind "
      + "a proxy is just a way to hide your real IP behind the IP of that proxy, but "
      + "there is no guarantee that the IP of that proxy won't be blocked as well.\n\n"
      + "The only truly reliable way to prevent IP blocks is rotating through a large "
      + "pool of residential IPs, by using a provider like Webshare which provides you "
      + "with a pool of >30M residential IPs (make sure to purchase "
      + '"Residential" proxies, NOT "Proxy Server" or "Static Residential"!).\n\n'
      + "You will find more information on how to easily integrate Webshare here: "
      + "https://github.com/chintamani-pala/yt-transcript-api"
      + "?tab=readme-ov-file#using-webshare"
   );
   static WITH_WEBSHARE_PROXY_CAUSE_MESSAGE = (
      "YouTube is blocking your requests, despite you using Webshare proxies. "
      + 'Please make sure that you have purchased "Residential" proxies and '
      + 'NOT "Proxy Server" or "Static Residential", as those won\'t work as '
      + 'reliably! The free tier also uses "Proxy Server" and will NOT work!\n\n'
      + 'The only reliable option is using "Residential" proxies (not "Static '
      + 'Residential"), as this allows you to rotate through a pool of over 30M IPs, '
      + "which means you will always find an IP that hasn't been blocked by YouTube "
      + "yet!\n\n"
   );

   constructor(videoId) {
      super(videoId);
      this._proxyConfig = null;
   }

   withProxyConfig(proxyConfig) {
      this._proxyConfig = proxyConfig;
      return this;
   }

   get cause() {
      // You must import these classes from your proxies.js
      const { WebshareProxyConfig, GenericProxyConfig } = require('./proxies');
      if (this._proxyConfig instanceof WebshareProxyConfig) {
         return RequestBlocked.WITH_WEBSHARE_PROXY_CAUSE_MESSAGE;
      }
      if (this._proxyConfig instanceof GenericProxyConfig) {
         return RequestBlocked.WITH_GENERIC_PROXY_CAUSE_MESSAGE;
      }
      return RequestBlocked.CAUSE_MESSAGE;
   }
}

class IpBlocked extends RequestBlocked {
   static CAUSE_MESSAGE = (
      RequestBlocked.BASE_CAUSE_MESSAGE
      + 'Ways to work around this are explained in the "Working around IP '
      + 'bans" section of the README (https://github.com/chintamani-pala/yt-transcript-api'
      + "?tab=readme-ov-file"
   );
   get cause() {
      return IpBlocked.CAUSE_MESSAGE;
   }
}

class TranscriptsDisabled extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "Subtitles are disabled for this video";
   get cause() {
      return TranscriptsDisabled.CAUSE_MESSAGE;
   }
}

class AgeRestricted extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = (
      "This video is age-restricted. Therefore, you are unable to retrieve "
      + "transcripts for it without authenticating yourself.\n\n"
      + "Unfortunately, Cookie Authentication is temporarily unsupported in "
      + "yt-transcript-api, as recent changes in YouTube's API broke the previous "
      + "implementation. I will do my best to re-implement it as soon as possible."
   );
   get cause() {
      return AgeRestricted.CAUSE_MESSAGE;
   }
}

class NotTranslatable extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "The requested language is not translatable";
   get cause() {
      return NotTranslatable.CAUSE_MESSAGE;
   }
}

class TranslationLanguageNotAvailable extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "The requested translation language is not available";
   get cause() {
      return TranslationLanguageNotAvailable.CAUSE_MESSAGE;
   }
}

class FailedToCreateConsentCookie extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = "Failed to automatically give consent to saving cookies";
   get cause() {
      return FailedToCreateConsentCookie.CAUSE_MESSAGE;
   }
}

class NoTranscriptFound extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = (
      "No transcripts were found for any of the requested language codes: {requested_language_codes}\n\n"
      + "{transcript_data}"
   );
   constructor(videoId, requestedLanguageCodes, transcriptData) {
      super(videoId);
      this._requestedLanguageCodes = requestedLanguageCodes;
      this._transcriptData = transcriptData;
   }
   get cause() {
      return NoTranscriptFound.CAUSE_MESSAGE
         .replace('{requested_language_codes}', this._requestedLanguageCodes)
         .replace('{transcript_data}', String(this._transcriptData));
   }
}

class PoTokenRequired extends CouldNotRetrieveTranscript {
   static CAUSE_MESSAGE = (
      "The requested video cannot be retrieved without a PO Token. If this happens, "
      + "please open a GitHub issue!"
   );
   get cause() {
      return PoTokenRequired.CAUSE_MESSAGE;
   }
}

module.exports = {
   YouTubeTranscriptApiException,
   CookieError,
   CookiePathInvalid,
   CookieInvalid,
   CouldNotRetrieveTranscript,
   YouTubeDataUnparsable,
   YouTubeRequestFailed,
   VideoUnplayable,
   VideoUnavailable,
   InvalidVideoId,
   RequestBlocked,
   IpBlocked,
   TranscriptsDisabled,
   AgeRestricted,
   NotTranslatable,
   TranslationLanguageNotAvailable,
   FailedToCreateConsentCookie,
   NoTranscriptFound,
   PoTokenRequired
};
